src/main/java/seedu/address/model/TaskMaster.java

```diff
@@ -72,21 +73,38 @@ public void setTasks(List<Task> tasks) {
         this.tasks.getInternalTaskList().clear();
         this.tasks.getInternalTaskList().addAll(tasks);
     }
-    
-    public void setComponents(List<TaskComponent> components) {
-    	this.tasks.getInternalComponentList().clear();
-    	this.tasks.getInternalComponentList().addAll(components);
-    }
 
     public void setTags(Collection<Tag> tags) {
         this.tags.getInternalList().setAll(tags);
     }
 
     public void resetData(Collection<? extends ReadOnlyTask> newTasks, Collection<? extends TaskComponent> newComponents, Collection<Tag> newTags) {
         setTasks(newTasks.stream().map(Task::new).collect(Collectors.toList()));
-        setComponents(newComponents.stream().map(TaskComponent::new).collect(Collectors.toList()));
+        rebuildComponentList();
         setTags(newTags);
     }
+    
+    /** Rebuilds the component list based on restored tasks, 
+     * as copy constructor of task component does not work.
+     */
+	public void rebuildComponentList() {
```

Comment:
I admit this method is quite messy and intuitive.
Yet the pointer of task component flies everywhere when calling the copy constructor.
The reference is not parallel but across with the original, and the newly created have not connection with each other.
If have time I think better to implement kind of association class.